core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -1450,12 +1463,67 @@ class GroupCoordinator(val brokerId: Int,
             group.maybeInvokeJoinCallback(member, joinResult)
             completeAndScheduleNextHeartbeatExpiration(group, member)
             member.isNew = false
+
+            group.addPendingSyncMember(member.memberId)
           }
+
+          val delayedSync = new DelayedSync(this, group, group.rebalanceTimeoutMs)
+          val groupKey = GroupKey(group.groupId)
+          syncPurgatory.tryCompleteElseWatch(delayedSync, Seq(groupKey))
         }
       }
     }
   }
 
+  private def maybeCompleteSyncExpiration(
+    group: GroupMetadata
+  ): Unit = {
+    val groupKey = GroupKey(group.groupId)
+    syncPurgatory.checkAndComplete(groupKey)
+  }
+
+  def tryCompletePendingSync(
+    group: GroupMetadata,
+    forceComplete: () => Boolean
+  ): Boolean = {
+    group.inLock {
+      group.currentState match {
+        case Dead | Empty | PreparingRebalance =>
+          forceComplete()
+        case CompletingRebalance | Stable =>
+          if (group.hasReceivedSyncFromAllMembers())
+            forceComplete()
+          else false
+      }
+    }
+  }
+
+  def onExpirePendingSync(
+    group: GroupMetadata
+  ): Unit = {
+    group.inLock {
+      group.currentState match {
+        case Dead | Empty | PreparingRebalance =>
+          debug(s"Received notification of sync expiration for group ${group.groupId} in ${group.currentState} state. Ignoring.")
```

Comment:
nit: We could log this debug string (except `ignoring`) on the top of the `onExpirePendingSync` method, so that we can see the log in each situation.

Reply:
I am not really convinced by this. My intent was to have a comment for the unexpected cases. I prefer to keep it as it is.

Reply:
SGTM. Thanks.